File: JULY-2025/10-07-2025/p3440.py
class Solution:
    def maxFreeTime(self, eventTime: int, startTime: List[int], endTime: List[int]) -> int:
        N = len(startTime)
        freeTimes = []
        eventLengths = []
        maxFreeTimes = []
        prevEnd = 0
        for i in range(N):
            curStart, curEnd = startTime[i], endTime[i]
            heapq.heappush(maxFreeTimes, curStart-prevEnd)
            if len(maxFreeTimes) > 3:
                heapq.heappop(maxFreeTimes)
            freeTimes.append(curStart-prevEnd)
            eventLengths.append(curEnd-curStart)
            prevEnd = curEnd
        freeTimes.append(eventTime-prevEnd)
        heapq.heappush(maxFreeTimes, eventTime-prevEnd)
        if len(maxFreeTimes) > 3:
            heapq.heappop(maxFreeTimes)
        def canRescheduleEventElsewhere(eventIdx):
            eventLength = eventLengths[eventIdx]
            free1, free2 = freeTimes[eventIdx], freeTimes[eventIdx+1]
            for time in maxFreeTimes:
                if time == free1 or time == free2:
                    if time == free1:
                        free1 = -1
                    elif time == free2:
                        free2 = -1
                elif time >= eventLength:
                    return True
            return False
                
        maxFreeTime = 0
        for i in range(N):
            if canRescheduleEventElsewhere(i):
                maxFreeTime = max(maxFreeTime, eventLengths[i]+freeTimes[i]+freeTimes[i+1])
            maxFreeTime = max(maxFreeTime, freeTimes[i]+freeTimes[i+1])
        
        return maxFreeTime

File: JULY-2025/10-07-2025/p3440.swift
class Solution {
  func maxFreeTime(_ eventTime: Int, _ startTime: [Int], _ endTime: [Int]) -> Int {
    let sizes = zip(endTime, startTime).map(-)
    let gaps = zip(chain(startTime, [eventTime]), chain([0], endTime)).map(-)
    let prefixSum = gaps.reductions(0,max)
    let suffixSum = Array(chain(gaps.dropFirst(2),[0]).reversed().reductions(max).reversed())
    let lrGaps = gaps.adjacentPairs().map(+)
    let meetingSizes = zip(sizes, zip(prefixSum, suffixSum).map(max)).map{$0<=$1 ? $0:0}
    return zip(lrGaps, meetingSizes).map(+).max()!
  }
}